frontend/utils/faviconBadge.js

```diff
@@ -0,0 +1,253 @@
+  if (typeof label === 'number' && label > 99 && options.abbreviate) {
+    label = abbreviateNumber(label)
+  }
+
+  // bubble needs to be larger for double digits
+  const len = (String(label)).length - 1
+  const width = options.width * r + (6 * r * len)
+  const height = options.height * r
+  // const top = size - height
+  // const left = size - width - r
+  const bottom = 16 * r
+  const right = 16 * r
+  // const radius = 2 * r
+  const arcRadius = Math.min(width, height) / 2
+
+  // webkit seems to render fonts lighter than firefox
+  context.font = (browser.webkit ? 'bold ' : '') + options.font
+  context.fillStyle = options.background
+  context.strokeStyle = options.background
+  context.lineWidth = r
+
+  // bubble
+  // draw circle bubble
+  context.arc(right - arcRadius, bottom - arcRadius, arcRadius, 0, 2 * Math.PI)
+
+  // draw rectangle bubble
+  // context.beginPath()
+  // context.moveTo(left + radius, top)
+  // context.quadraticCurveTo(left, top, left, top + radius)
+  // context.lineTo(left, bottom - radius)
+  // context.quadraticCurveTo(left, bottom, left + radius, bottom)
+  // context.lineTo(right - radius, bottom)
+  // context.quadraticCurveTo(right, bottom, right, bottom - radius)
+  // context.lineTo(right, top + radius)
+  // context.quadraticCurveTo(right, top, right - radius, top)
+  // context.closePath()
+  context.fill()
+
+  // bottom shadow
+  // context.beginPath()
+  // context.strokeStyle = 'rgba(0,0,0,0.3)'
+  // context.moveTo(left + radius / 2.0, bottom)
+  // context.lineTo(right - radius / 2.0, bottom)
+  // context.stroke()
+
+  // label
+  context.fillStyle = options.color
+  context.textAlign = 'right'
+  context.textBaseline = 'top'
+
+  // unfortunately webkit/mozilla are a pixel different in text positioning
+  // NOTE: temporarily comment this drawing text because we don't need it
+  // context.fillText(label, r === 2 ? 29 : 15, browser.mozilla ? 7 * r : 6 * r)
+}
+
+const refreshFavicon = () => {
+  // check support
+  if (getCanvas() instanceof HTMLCanvasElement) {
+    setFaviconTag(getCanvas().toDataURL())
+  }
+}
+
+const abbreviateNumber = (label): string => {
+  const metricPrefixes = [
+    ['G', 1000000000],
+    ['M', 1000000],
+    ['k', 1000]
+  ]
+
+  for (let i = 0; i < metricPrefixes.length; ++i) {
+    if (label >= metricPrefixes[i][1]) {
+      return round(label / metricPrefixes[i][1]) + metricPrefixes[i][0]
+    }
+  }
+
+  return String(label)
+}
+
+const round = (value, precision): string => {
+  return Number(value).toFixed(precision)
+}
+
+const FaviconBadge = {
+  setOptions: function (custom: Object) {
+    options = {}
+
+    for (const key in defaults) {
+      options[key] = custom[key] || defaults[key]
+    }
+  },
+  setImage: function (url: string) {
+    currentFavicon = url
+    refreshFavicon()
+  },
+  setBubble: function (label: string | number, color?: string) {
+    drawFavicon(String(label || ''), color)
```

Comment:
is there a reason why `drawFavicon` has to be wrapped in `setBubble`? it looks like `drawFavicon` is only used in here. can this just be `FaviconBadge.drawFavicon(this.shouldSetBadge ? 1 : 0)` instead for better readability?

Reply:
In terms of workflow the function name should be `drawFavicon`, but for the user's point of view, it should be `setBubble`.
I like to keep the function name `setBubble` if it's not a big deal.